dbt/tracking.py

```diff
@@ -56,21 +59,20 @@ def initialize(self):
         tracker.set_subject(subject)
 
     def set_cookie(self):
-        cookie_dir = os.path.dirname(COOKIE_PATH)
         user = {"id": str(uuid.uuid4())}
 
-        dbt.clients.system.make_directory(cookie_dir)
+        dbt.clients.system.make_directory(self.cookie_dir)
 
-        with open(COOKIE_PATH, "w") as fh:
+        with open(self.cookie_path, "w") as fh:
             yaml.dump(user, fh)
 
         return user
 
     def get_cookie(self):
-        if not os.path.isfile(COOKIE_PATH):
+        if not os.path.isfile(self.cookie_path):
```

Comment:
is it possible to defer the creation of this cookie until the profile is loaded? With the current implementation, dbt will create a cookie if it doesn't exist in the specified profile directory. That means if you errantly specify an incorrect profile dir, dbt will litter .user.yml files all over the place. This won't affect tracking and i don't think it will cause any problems for the end-user, but it's not so polite :)

I figure we can only create the cookie if dbt is able to find and load the profiles.yml file. To be sure: I tested this out, and dbt does _not_ send events if the profiles.yml file is unloadable, which is good. This is mostly a cosmetic/tidyness comment

Reply:
It's possible, but not easy - I would estimate it at a few days of work, as a ton of main.py would have to be refactored. We start doing tracking things before we load a profile - really before we ever even start inspecting which command was passed, so we don't even know if we _will_ load a profile! And sometimes, in order to load a profile we actually need to load the project first so we can figure out the profile name, etc... Accordingly, per our slack conversation, I'm going to merge this as-is - we should get this fix in to grace-kelly, even if it's a bit less than optimal.

I am very enthusiastic about the idea of refactoring main.py to support this behavior - currently even answering the question "does this command require profile, project, both or neither" requires checking multiple functions and carefully walking through behavior, and I don't like that. We could also handle the issues raised in #1189 when/if we do that.

Reply:
totally agree @beckjake, thanks for the color. Ship it!